File: main/subsets/subsets.cs
public class Solution {
    public IList<IList<int>> Subsets(int[] nums) {
        IEnumerable<List<int>> SubsetsTo(int stop) {
            if (stop == 0) {
                yield return new List<int>();
                yield break;
            }

            var element = nums[stop - 1];

            foreach (var withoutElement in SubsetsTo(stop - 1)) {
                yield return withoutElement;

                var withElement = new List<int>(withoutElement.Count + 1);
                withElement.AddRange(withoutElement);
                withElement.Add(element);
                yield return withElement;
            }
        }

        return SubsetsTo(nums.Length).Select(subset => (IList<int>)subset)
                                     .ToList();
    }
}

File: main/subsets/subsets.cpp
class Solution {
public:
    static vector<vector<int>> subsets(const vector<int>& nums);
};

vector<vector<int>> Solution::subsets(const vector<int>& nums)
{
    vector<vector<int>> powerset;
    powerset.reserve(size_t{1} << nums.size());
    powerset.emplace_back(); // the empty subset

    for (const auto element : nums) {
        const auto old_size = powerset.size();
        for (size_t i {0}; i != old_size; ++i) {
            powerset.push_back(powerset[i]);
            powerset.back().push_back(element);
        }
    }

    return powerset;
}
